### ISSUE TITLE
angular utilities oder patterns oder sonstwas....

### ISSUE THREAD

**Issue by dherges** [spektrakel-blog/sparkles#165]
```ts
import { Component, Input, ViewChild, ElementRef, ComponentFactoryResolver, ApplicationRef, Injector } from '@angular/core';
import { CdkPortal, PortalOutlet, DomPortalOutlet } from '@angular/cdk/portal';
import { Overlay, OverlayRef } from '@angular/cdk/overlay';

/**
 * `<u-fullscreen>` is an utility that beams out arbitrary content to an overlay panel covering the full size of the browser screen.
 *
 * ##### How To Use
 *
 * Embedd arbitrary content to the component and toggle the display mode through an input-bound property:
 *
 * ```html
 * <u-fullscreen [uIsFullScreen]="true|false">
 *   <h1>put your custom content here</h1>
 * </u-fullscreen>
 * ```
 *
 * ##### How It Works
 *
 * Internally, it uses [Angular CDK Portals](https://material.angular.io/cdk/portal/api) to capture the projected content and beam it to a `PortalOutlet`.
 *
 * When in fullscreen mode, the portal is beamed to an [Angular CDK Overlay](https://material.angular.io/cdk/overlay/overview).
 * Otherwise, the portal is beamed to the host element (appended as child to `<u-fullscreen>`).
 *
 * @experimental
 */
@Component({
  selector: 'u-fullscreen',
  template: `
    <ng-template cdkPortal>
      <div [class.fullscreen]="uIsFullScreen">
        <ng-content></ng-content>
      </div>
    </ng-template>
  `,
  styles: [`
    .fullscreen {
      position: fixed;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      z-index: 1050;
      background-color: white;
    }
  `]
})
export class FullScreenComponent {

  @Input()
  public uIsFullScreen = false;

  /** @internal */
  @ViewChild(CdkPortal)
  public portal: CdkPortal;

  private outlet: PortalOutlet;

  constructor(
    private element: ElementRef,
    private overlay: Overlay,
    private cfr: ComponentFactoryResolver,
    private appRef: ApplicationRef,
    private injector: Injector
  ) {}

  ngOnChanges(changes) {
    if (changes['uIsFullScreen']) {
      this.toggle();
    }
  }

  toggle() {
    if (this.outlet) {
      this.outlet.detach();
    }

    if (this.uIsFullScreen) {
      this.outlet = this.overlay.create();
    } else {
      this.outlet = new DomPortalOutlet(this.element.nativeElement, this.cfr, this.appRef, this.injector);
    }

    this.outlet.attach(this.portal);
  }

}
```

**Comment by dherges** [comment]
### Touch Events and NgZone
 
```
  constructor(
    private _cdr: ChangeDetectorRef,
    public _sliderRef: ElementRef,
    private ngZone: NgZone,
    private renderer: Renderer2
  ) {
    const nativeElement = this._sliderRef.nativeElement;
    this.ngZone.runOutsideAngular(() => {
      this.hammerSliderEl = new Hammer(nativeElement);
      this.hammerSliderEl.on('panmove', guardOutsideAngular(event => this.onPanMove(event)));
    });
  }
 
  ngOnDestroy() {
    if (this.hammerSliderEl) {
      // Gracefully clean up DOM Event listeners that were registed by hammer.js
      this.hammerSliderEl.off('panmove');
      this.hammerSliderEl.stop(true);
      this.hammerSliderEl.destroy();
    }
  }
 
  onPanMove($event) {
     if (/* … magic expression … */) {
       /* … only internal changes, no rendering needed … */)
     } else {
        this._cdr.detectChanges();
     }
  }
 ```
 


```
    // Run in ngZone because it should update slider values outside component
    this.unregisterListeners = [
      this.renderer.listen(nativeElement, 'touchstart', event => this._onTouchStart(event)),
      this.renderer.listen(nativeElement, 'mousedown', event => this._onClick(event))
    ];
 
    if (this.unregisterListeners) {
      this.unregisterListeners.forEach(fn => fn());
    }
```

**Comment by dherges** [comment]
### DOM Outlet

States
 - beamed (attached to arbitrary element)
 - beamed to fullscreen (attached to body)
 - in place: attached to parent element
 - hidden (not visible)


```ts
import { Component, Input, ViewChild, ElementRef, ComponentFactoryResolver, ApplicationRef, Injector } from '@angular/core';
import { CdkPortal, PortalOutlet, DomPortalOutlet } from '@angular/cdk/portal';
import { Overlay, OverlayRef } from '@angular/cdk/overlay';

/**
 * `<sp-fullscreen>` is an utility that beams out arbitrary content to an overlay panel covering the full size of the browser screen.
 *
 * ##### How To Use
 *
 * Embed arbitrary content to the component and toggle the display mode through an input-bound property:
 *
 * ```html
 * <sp-fullscreen [spIsFullScreen]="true|false">
 *   <h1>put your custom content here</h1>
 * </sp-fullscreen>
 * ```
 *
 * ##### How It Works
 *
 * Internally, it uses [Angular CDK Portals](https://material.angular.io/cdk/portal/api) to capture the projected content and beam it to a `PortalOutlet`.
 *
 * When in fullscreen mode, the portal is beamed to an [Angular CDK Overlay](https://material.angular.io/cdk/overlay/overview).
 * Otherwise, the portal is beamed to the host element (appended as child to `<sp-fullscreen>`).
 *
 * @experimental
 */
@Component({
  selector: 'sp-fullscreen',
  template: `
    <ng-template cdkPortal>
      <div [class.fullscreen]="spIsFullScreen">
        <ng-content></ng-content>
      </div>
    </ng-template>
  `,
  styles: [`
    .fullscreen {
      position: absolute;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      z-index: 1050;
      background-color: white;
    }
  `]
})
export class FullScreenComponent {

  @Input()
  public spIsFullScreen = false;

  /** @internal */
  @ViewChild(CdkPortal)
  public portal: CdkPortal;

  private outlet: PortalOutlet;

  constructor(
    private element: ElementRef,
    private overlay: Overlay,
    private cfr: ComponentFactoryResolver,
    private appRef: ApplicationRef,
    private injector: Injector
  ) {}

  ngOnChanges(changes) {
    if (changes['spIsFullScreen']) {
      this.toggle();
    }
  }

  toggle() {
    if (this.outlet) {
      this.outlet.detach();
    }

    if (this.spIsFullScreen) {
      this.outlet = this.overlay.create();
    } else {
      this.outlet = new DomPortalOutlet(this.element.nativeElement, this.cfr, this.appRef, this.injector);
    }

    this.outlet.attach(this.portal);
  }

}
```

### Content Projectoon: a custom lifecycle hook

Content Projection / View Children: custom lifecycle hooks

```ts 
/**
* Custom lifecycle hook that is invoked when a `QueryList` property, annotated with `@ContentChildren()`,
* of the parent component receives changes from dynamic content.
*
* It should be used in combination with the `@ContentChildrenChanges()` and `@ContentChildren()` decorators.
* Minimal usage example:
*
* ```ts
* @Component({
*   template: `<ng-content></ng-content>`
* })
* export class FooComponent implements ContentChanges {
*
*   @ContentChildren(BarComponent)
*   @ContentChildrenChanges()
*   barComponents: QueryList<BarComponent>;
*
*   // Called when a `QueryList` changes
*   myOnContentChanges(change) {
*     // Reflects the value of `this.barComponents`
*     const changesToBarComponents = change['barComponents'];
*   }
* }
* ```
*
* @stable
*/
export interface ContentChanges {
  myOnContentChanges(change: QueryListChange);
}
```

#### Going Fullscreen

Use browser [Fullscreen API](https://developer.mozilla.org/en-US/docs/Web/API/Fullscreen_API)

```ts
import { DOCUMENT } from '@angular/common';

@Directive({
  selector: 'spFullscreen'
})
export class FullscreenDirective {

  @Input()
  public spFullscreen = false;

  constructor(
    private element: ElementRef,
    @Inject(DOCUMENT) @Optional() private document: Document
  ) {}

  public show() {
    if (this.element.nativeElement && typeof this.element.nativeElement.requestFullscreen === 'function') {
      return this.element.nativeElement.requestFullscreen();
    } else {
      return Promise.resolve();
    }
  }

  public hide() {
    if (this.document) {
      return this.document.exitFullscreen();
    } else {
      return Promise.resolve();
    }
  }

}
```

**Comment by dherges** [comment]
### An Approach for Theming With Shadow DOM And Web Components

Idea: in Shadow DOM environments, let's go for CSS Variables! The component gives hooks how it can be styled, the container app decides what it looks like.

https://developers.google.com/web/fundamentals/web-components/shadowdom#stylefromoutside

```scss 
@function my-theme-color($key: "primary”) {
  // Returns the simple CSS variable expression `var(--primary)`
  @return var(--#{$key});
}
 
@function my-theme-color($key: "primary”, $level: 0, $opacity: 0) {
  // TODO: return HSLA calculation w/ CSS variable expression
}
```

**Comment by dherges** [comment]
### DOM Outlet

Use CDK [Portal API](https://material.angular.io/cdk/portal/overview) to attach content to another physical location in the DOM.


### Content Projection: a custom lifecycle hook

Become notified about changes in [ViewChildren](https://angular.io/api/core/ViewChildren), [ContentChildren](https://angular.io/api/core/ContentChildren). Implement a decorator to subscribe `changes` observable.


### Fullscreen API

Use browser [Fullscreen API](https://developer.mozilla.org/en-US/docs/Web/API/Fullscreen_API) to switch into fullscreen display.

**Comment by dherges** [comment]
### WebRTC

https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/enumerateDevices

### QR Codes

https://github.com/zxing-js/ngx-scanner

**Comment by dherges** [comment]
Ich möchte Dir mit diesem Post Rückmeldung und Ideen zum "Advanced Component Design" geben:

Baut auf/als Erweiterung zu [99_Component_Design.pptx](../tree/master/ent/01_folien/99_Component_Design.pptx)

### Design Systems on Angular

Metapher: Automobil

Building Blocks for a car: mirror, wheel, exhaust, ...

The product: Golf VI, Golf VII, Golf VII Variant, Passat, Polo, ...

Can we re-use the same mirror in different car models?

Maybe, yes.

#### Context

https://unsplash.com/photos/A53o1drQS2k

Context: the car model

#### Variations

https://unsplash.com/photos/JFQE8Ed2pIg

Variations: the exhaust needs to work on different car models, thus we need variations of an exhaust

Would you build a dedicated exhaust for each car model? Definetely not. We'd build an exhaust that fits into and adapts to the context.

#### Component

https://unsplash.com/photos/qWwpHwip31M

Actually, the component is not the exhaust, but the moulding blank ("Rohling") of an exhaust. 

Or: the component is a set of building instructions for the exhaust.

Other familiar case studies: web shop and shopping cart:
 - shopping cart preview icon
 - shopping cart dropdown
 - full-page shopping cart to increase/decrease quantities
 - read-only shopping cart prior to payment confirmation.

#### Cohesion, Coupling

The questions: one shopping cart to rule them all?

 - x axis: number of contexts
 - y axis: number of variations
 - lower left: one component
 - upper right: many components
 - the difficult parts: anything in between

Ideal goal: high cohesion, low coupling

Button and a Button Toolbar: the button needs to work as part of a toolbar (high cohesion between toolbar and button). The button also needs to work on its own (loose coupling between toolbar and button).


#### A Button...

```html
<button myButton>Go ahead</button>
```

```html
<button myButton="primary">Go ahead</button>
```

#### ...and a Button Toolbar

```html
<button-toolbar>
  <button myButton>List</button>
  <button myButton>Grid</button>
</button-toolbar>
```

What if...
 - a button inside toolbar should always be a "primary" button?
 - each button can be borderless or bordered (e.g. two variations: style and border)? 

```html
<toolbar>
  <toolbar-button>List</toolbar-button>
  <toolbar-button>Grid</toolbar-button>
</toolbar>
```

```html
<button myButton="primary-bordered">Go ahead</button>
<button myButton="primary" [myBordered]="true">Go ahead</button>
<button myButton class="btn-primary btn-bordered">Go ahead</button>
```

#### A Dumb Idea?

```html
<primary-btn>Go ahead</primary-btn>
<secondary-btn>Go back</secondary-btn>
```

=> lots of possibilties. there may be a golden gun, but there's no silver bullet


#### Open/Closed Principle

Von...

```ts
@Directive({
  selector: 'button'
})
export class ButtonDirective {}
```

...zu

```ts
@Directive({
  selector: 'button[myButton]'
})
export class ButtonDirective {
  @Input() myButton: 'primary' | 'secondary' = 'primary';
}
```

...zu

```ts
@Directive({
  selector: '[myButton]'
})
export class ButtonDirective {
  @Input() myButton: 'primary' | 'secondary' = 'primary';
}
```


#### First Live Coding

(tbd)








### Component vs. Directive

Case Study: `<my-button>` vs. `<button myButton>`

Take Aways:

Komponente
 - custom element
 - native Events wie click, select, change müssen ggf. "nach-implementiert" werden
 - bringt ein HTML Template und Component Stylesheets mit
 - kann optional mit `<ng-content>` Kind-Inhalte projezieren
 - hat optional ViewChildren und optional ContentChildren

Direktive
 - kann an beliebiges Element attached werden
 - Native Events sind "out-of-the-box" verfügbar.
 - implizite Content Projection
 - hat optional ContentChildren, aber keine ViewChildren

Mentales Model: Komponente = Direktive + Template


### Content Projection

Mit Selektoren

```ts
import { ButtonDirective } from '@my/components/button';

@Component({
  selector: 'my-button-toolbar',
  template: `<ng-content select="button[myButton]"></ng-content>`
})
export class ButtonToolbarComponent {

  @ContentChildren(ButtonDirective)
  public buttons$: QueryList<ButtonDirective>;
}
```

##### Dynamischer Content

```html
<my-button-toolbar>
  <button myButton *ngFor="let button of allButtons">{{ button.label }}</button>
</my-button-toolbar>
<hr>
<button myButton class="meta-button" (click)="onAddAnotherButton()">Button in Toolbar hinzufügen</button>
```

Lösungs-Idee: [`QueryList.changes: Observable<any>`](https://angular.io/api/core/QueryList#changes)

```ts
@Component({
  selector: 'my-button-toolbar',
  template: `<ng-content select="button[myButton]"></ng-content>`
})
export class ButtonToolbarComponent implements AfterContentInit, OnDestroy {

  @ContentChildren(ButtonDirective)
  public buttons$: QueryList<ButtonDirective>;

  private buttonContent: Subscription;

  ngAfterContentInit() {
    this.buttonContent = this.button$.changes.subscribe(
      childButtons => { /* react to dynamic content change */ }
    );
  }

  ngOnDestroy() {
    if (thus.buttonContent) {
      this.buttonContent.unscubscribe();
    }
  }
}
```

##### ngOnContentChanges()

Syntax Sugar for a `ngOnContentChanges()` lifecycle hook?

```ts
@Component({
  selector: 'my-button-toolbar',
  template: `<ng-content select="button[myButton]"></ng-content>`
})
export class ButtonToolbarComponent implements ContentChanges{

  @ContentChildren(ButtonDirective)
  public buttons$: QueryList<ButtonDirective>;

  ngOnContentChanges(change) {
    const buttonChange = change['button$'];
    /* react to dynamic content change... */
  }
}
```

Live Coding: [implement a custom lifecycle hook with TypeScript decorator](https://github.com/angular/angular/issues/25016)

### Eine Alternative zu Content Projection: `<ng-template>`

Idee: [Creating elements from a `<template>` (Shadow DOM)](https://developers.google.com/web/fundamentals/web-components/customelements#fromtemplate)

Angular Equivalent: [`<ng-template>`](https://angular.io/api/common/NgTemplateOutlet#example)


##### Flavour 1: ng-template als Input property

Anwendung:

```html
<ng-template #myTemplate let-button>
  <span>Hello {{ button.label }}</span>
</ng-template>
<my-button-toolbar buttonTemplate="myTemplate" buttons="buttons"></my-button-toolbar>
```

Implementierung:

```ts
@Component({
  selector: 'my-button-toolbar',
  template: `
    <button myButton *ngFor="let button of buttons">
      <ng-container *ngTemplateOutlet="buttonTemplate; context: button"></ng-container>
    </button>
  `
})
export class ButtonToolbarComponent {
  @Input()
  public buttons: ButtonInterface[];

  @Input()
  public buttonTemplate: TemplateRef<any>;
}
```

##### Flavour 2: ng-template als content children

Anwendung:

```html
<my-button-toolbar buttons="buttons">
  <ng-template buttonTemplate let-button>
    <span>Hello {{ button.label }}</span>
  </ng-template>
</my-button-toolbar>
```

Implementierung:
```ts
@Component({
  selector: 'my-button-toolbar',
  template: `
    <button myButton *ngFor="let button of buttons">
      <ng-container *ngTemplateOutlet="buttonTemplate.templateRef; context: button"></ng-container>
    </button>
  `
})
export class ButtonToolbarComponent {
  @Input()
  public buttons: ButtonInterface[];

  @ContentChild(ButtonTemplateDirective)
  public buttonTemplate: ButtonTemplateDirective;
}

@Directive({ selector: '[buttonTemplate]' })
export class ButtonTemplateDirective {
  constructor(
    public templateRef: TemplateRef<any>
  ) {}
}
```



### Performance Optimization

Touch Events and NgZone

Herleitung: property binding, event binding, banana-in-the-box-syntax als Reminder

Refresher/Reminder: event bubbling from children to parent -> triggers change detection

Mit zone.js und NgZone: jedes DOM event fragt eine Change Detection an

Spezialfall Touch Gesten: eine Geste besteht aus mehreren Events, eine Geste sollte eine Change Detection anfragen. Jedoch nicht jedes Event.

```ts
  constructor(
    private cdr: ChangeDetectorRef,
    private element: ElementRef,
    private ngZone: NgZone,
    private renderer: Renderer2
  ) {
    const nativeElement = this.element.nativeElement;
    this.ngZone.runOutsideAngular(() => {
      this.removeListener = this.renderer.listen('touchmove', evt => this.onTouchMove(evt));
    });
  }
 
  onTouchMove(evt) {
     if (/* … detecting pinch-to-zoom-gesture … */) {
       /* compute whether touch events are pinch-to-zoom */)
     } else {
       const scaleFactor = /* pinch-to-zoom detected! */
       this.updateScaledView(scaleFactor);
       this.cdr.markForCheck();
     }

  }

  ngOnDestroy() {
    this.removeListener();
  }

}
```

Fallstricke:

```ts
  @Output() fooChange = new EventEmitter();

  /** WATCH OUT: this method is not running in NgZone! */
  onTouchMove() {
    this.fooChange.next(); // <-- target of event binding would also run outside NgZone
  }
```

===>

```ts
  /** WATCH OUT: this method is not running in NgZone! */
  onTouchMove() {
    this.ngZone.run(() => {
      this.fooChange.next(); // <-- target of event binding will execute in NgZone (as expected)
    });
  }
```

```ts
  ngOnDestroy() {
    this.removeListener();
  }
```


---
